lib/benchee/benchmark/runner.ex

```diff
@@ -0,0 +1,154 @@
+                   config: %Configuration{warmup: warmup}
+                 }) do
+    warmup_context = %ScenarioContext{scenario_context | show_fast_warning: false,
+                                                         run_time: warmup}
+    measure_runtimes(scenario, warmup_context)
+  end
+
+  def run_benchmark(scenario, scenario_context = %ScenarioContext{
+                      config: %Configuration{time: run_time, print: %{fast_warning: fast_warning}}
+                    }) do
+    measurement_context =
+      %ScenarioContext{scenario_context | show_fast_warning: fast_warning,
+                                          run_time: run_time}
+    measure_runtimes(scenario, measurement_context)
+  end
+
+  defp pmap(collection, func) do
+    collection
+    |> Enum.map(fn(_) -> Task.async(func) end)
+    |> Enum.map(&Task.await(&1, :infinity))
+    |> List.flatten
+  end
+
+  defp measure_runtimes(scenario, %ScenarioContext{run_time: 0}), do: scenario
+  defp measure_runtimes(scenario, scenario_context = %ScenarioContext{run_time: run_time}) do
+    end_time = current_time() + run_time
+    :erlang.garbage_collect
+    {num_iterations, initial_run_time} =
+      determine_n_times(scenario, scenario_context)
+    updated_scenario = %Scenario{scenario | run_times: [initial_run_time]}
+    new_context =
+      %ScenarioContext{scenario_context | current_time: current_time(),
+                                          end_time: end_time,
+                                          num_iterations: num_iterations}
+    do_benchmark(updated_scenario, new_context)
+  end
+
+  defp current_time do
+    :erlang.system_time :micro_seconds
+  end
+
+  # If a function executes way too fast measurements are too unreliable and
+  # with too high variance. Therefore determine an n how often it should be
+  # executed in the measurement cycle.
+  @minimum_execution_time 10
+  @times_multiplicator 10
+  defp determine_n_times(scenario, scenario_context = %ScenarioContext{
+                           show_fast_warning: fast_warning, printer: printer
+                         }) do
+    run_time = measure_call(scenario, scenario_context)
+    if run_time >= @minimum_execution_time do
+      {1, run_time}
+    else
+      if fast_warning, do: printer.fast_warning()
+      new_context =
+        %ScenarioContext{scenario_context | num_iterations: @times_multiplicator}
+      try_n_times(scenario, new_context)
+    end
+  end
+
+  defp try_n_times(scenario, scenario_context = %ScenarioContext{
+                     num_iterations: num_iterations
+                   }) do
+    run_time = measure_call_n_times(scenario, scenario_context)
+    if run_time >= @minimum_execution_time do
+      {num_iterations, run_time / num_iterations}
+    else
+      new_context = %ScenarioContext{
+        scenario_context | num_iterations: num_iterations * @times_multiplicator
+      }
+      try_n_times(scenario, new_context)
+    end
+  end
```

Comment:
I still look at this and I wanna make `try_n_times` and `determine_n_times` one. Not here, alas.